### ISSUE TITLE
consider using expand_selection in place of existing select_link function

### ISSUE THREAD

**Issue by toolboxen** [renerocksai/sublime_zk#2]
I noticed in function of `def select_link(self)` that it might be possible to use a command already provided by sublime text of `expand_selection`:

> expand_selection
    Extends the selection up to predefined limits.
        to [Enum]: Values: bol, hardbol, eol, hardeol, bof, eof, brackets, line, tag, scope, indentation.
source: http://docs.sublimetext.info/en/latest/reference/commands.html

When cursor is placed on word, calling `expand_selection` with arg of `brackets` expands selection up to first matching pair of brackets. For your plugin it would expand selection to get all of note ID up to first instance of `[]`.

With this method, users would be able to use both double brackets `[[]]` and single brackets `[]` for note id's. I know that some different zettelkasten users like to use both single and double brackets for denoting zettel links.

**Comment by renerocksai** [comment]
Another great idea. This is definitely something I'm going to look into when I get back from work :)

**Comment by toolboxen** [comment]
If you go this route, you might need to re-work how the plugin does the brackets. It's hard-coded right now to look for double brackets but if you use `expand_selection` it would allow users to use single brackets while simplifying your code a little since the `expand_selection` command extends the selection just to the first bracket.

When I wrote https://github.com/Zettelkasten-Method/atom-wikilink I allowed users to select their own brackets and used a similar method to `expand_selection` to just grab whatever was inside the brackets.

**Comment by renerocksai** [comment]
Thanks for the hint! I was aware that supporting different bracket styles for links (thinking of only single and double quare brackets, though) would mean to change the bracket lookup. It's likely possible to just do a regexp search instead of looking for fixed strings.  `expand_selection`, as you imply, would focus on the text between brackets, right. OK, fair enough, in that case I can always just extend the selection by one to the left and right in case there's more brackets.

So in short: I'm sure it can be worked out without too much hassle :)

Thanks also for providing a link into the atom-wikilink, I'll check it out "for inspiration" :)

**Comment by toolboxen** [comment]
Sure. As an side: If you look at how @divinedominion is doing the linking in the archive app he just highlights the text between the two brackets as the brackets are just markers for the link itself. same thing in Nvalt, the link highlighting is just a line underneath the text, the brackets on either side are not underlined. In atom-wikilink I was even more minimalist enough as I consider the brackets to be markers enough without the highlighting. The reason for this is that highlighting was expensive in Atom when opening files with lots of links, it tooks a long time to parse the file and apply the highlighting to every link. Highlighting in sublime is a lot faster but keep in mind that generally speaking regex approaches to highlighting are more expensive than scoped approaches.

**Comment by renerocksai** [comment]
I had highlighting just between the brackets initially. I did this by
offsetting the range from [[ until ]] by +/-2. I changed it because the
highlighting of footnote references (I think by MarkdownEditing)
interfered, even my footnote syntax interfered: there is always a
[something] inside a [[something]]. Writing this, I think I can tweak the
footnote regexp to exclude leading opening and closing brackets to a string
in brackets; then I can re-introduce the offset to the note links.

Now, I am not sure how to do the highlighting with just markers. Can you
place markers so that sublime would recognize them as link markers? I have
two issues with that: 1. if I set the scope of the note ID to
something.link, syntax coloring would color them the same as other, like
http links. I tried that. Made my note links underlined green, didn't like
it. So I chose a different scope which I find quite pleasing, especially
since it distinguishes note links from external-to-archive links. 2. I have
no clue how to make Sublime Text recognize new markers that create a
(optional: special note-) link scope, apart from creating a syntax file --
like I did with the footnotes. Such syntax file - caused highlighting might
be quicker than processing the text a second time in my plugin, applying
artificial "underline" scopes and have Sublime-Text render the view.... Am
I missing something?

**Comment by toolboxen** [comment]
I see what you're saying. At the moment, I don't know enough about the specifics of how Sublime handles the highlighting. So I'm not sure if there's a better way to handle it then the method you're currently using. Atom actually has a cleaner approach to highlighting, even it if it is slower than Sublime 😄

**Comment by renerocksai** [comment]
OK a couple of things: I thought we don't want custom syntaxes anymore? :-)

To achieve [[links]] and [links] I basically simply changed my regex and search strings. 

Since note-link highlighting is done in code and doesn't seem to be a problem in terms of performance (it uses sublime's regex-search, applies the scopes, and does some bookkeeping), I really tend to leave it like it is.

I'll have to update the screenshot. Oh, and we need a setting for whether link insertion should use single or double brackets now :)

**Comment by renerocksai** [comment]
[single] and [[double]] work now. so i will close this issue now. feel free to re-open if you're not happy :)
